### PR TITLE
Implement runtime related flows for automatic scenario assignments

### PR DESCRIPTION
**Pull request by kfurgol**
**Description**

Changes proposed in this pull request:

- Added logic which handles scenarios during:
  - runtime creation
  - runtime update
  - setting label on runtime with `setRuntimeLabel` mutation
  - deleting label on runtime with `deleteRuntimeLabel` mutation
- Added unit tests for new logic and some missing ones

**Related issue(s)**
#1031

### CODE REVIEW

File: components/director/internal/domain/runtime/service.go [hunk 1]
```diff
@@ -288,5 +335,70 @@ func (s *service) DeleteLabel(ctx context.Context, runtimeID string, key string)
 		return errors.Wrapf(err, "while deleting Runtime label")
 	}
 
+	if key == model.ScenariosKey {
```
Comment (aszecowka) [inline components/director/internal/domain/runtime/service.go:338]: the same as comments above

File: components/director/internal/domain/runtime/service.go [hunk 2]
```diff
@@ -147,6 +164,13 @@ func (s *service) Create(ctx context.Context, in model.RuntimeInput) (string, er
 		in.Labels[model.ScenariosKey] = model.ScenariosDefaultValue
 	}
 
+	scenarios, err := s.synchroniseScenariosWithAssignments(ctx, in.Labels)
```
Comment (pkosiec) [inline components/director/internal/domain/runtime/service.go:167]: As Adam pointed out in his docs, we probably shouldn't keep DEFAULT scenario on the list

File: components/director/internal/domain/scenarioassignment/engine.go [hunk 3]
```diff
@@ -28,3 +38,45 @@ func (engine) RemoveAssignedScenarios(in []*model.AutomaticScenarioAssignment) e
 	// remove scenarios from runtimes, which have label matching selector
 	return nil
 }
+
+func (s engine) GetScenariosForSelectorLabels(ctx context.Context, labels map[string]interface{}) ([]string, error) {
```
Comment (pkosiec) [inline components/director/internal/domain/scenarioassignment/engine.go:42]: Hmm, what do you think about the following name `ListScenariosFromSelectorLabels` or `ComputeScenariosFromSelectorLabels`. I think it's more clear what it does.

However, for me the issue is also in the `SelectorLabels` name.  I needed to read the method to know what it does. But unfortunatelyI don't have better idea for that.

File: components/director/internal/domain/runtime/service.go [hunk 4]
```diff
@@ -278,10 +366,127 @@ func (s *service) DeleteLabel(ctx context.Context, runtimeID string, key string)
...
+	}
+
 	err = s.labelRepo.Delete(ctx, rtmTenant, model.RuntimeLabelableObject, runtimeID, key)
 	if err != nil {
 		return errors.Wrapf(err, "while deleting Runtime label")
 	}
 
+	finalScenarios := make([]interface{}, 0)
+
+	if key == model.ScenariosKey {
```
Comment (pkosiec) [inline components/director/internal/domain/runtime/service.go:381]: This is our final work of refactoring Set and Delete Label methods:
```go
func (s *service) SetLabel(ctx context.Context, labelInput *model.LabelInput) error {
	rtmTenant, err := tenant.LoadFromContext(ctx)
	if err != nil {
		return errors.Wrapf(err, "while loading tenant from context")
	}

	err = s.ensureRuntimeExists(ctx, rtmTenant, labelInput.ObjectID)
	if err != nil {
		return err
	}

	currentRuntimeLabels, err := s.getCurrentLabelsForRuntime(ctx, rtmTenant, labelInput.ObjectID)
	if err != nil {
		return err
	}

	newRuntimeLabels :=  currentRuntimeLabels // TODO: copy it; this are labels after modification
	newRuntimeLabels[labelInput.Key] = labelInput.Value

	err = s.upsertScenariosLabelIfShould(ctx, labelInput.ObjectID, labelInput.Key, currentRuntimeLabels, newRuntimeLabels)
	if err != nil {
		return err
	}

	if labelInput.Key != model.ScenariosKey {
		err = s.labelUpsertService.UpsertLabel(ctx, rtmTenant, labelInput)
		if err != nil {
			return errors.Wrapf(err, "while creating label for Runtime")
		}
	}

	return nil
}


func (s *service) DeleteLabel(ctx context.Context, runtimeID string, key string) error {
	rtmTenant, err := tenant.LoadFromContext(ctx)
	if err != nil {
		return errors.Wrapf(err, "while loading tenant from context")
	}

	err = s.ensureRuntimeExists(ctx, rtmTenant, runtimeID)
	if err != nil {
		return err
	}

	currentRuntimeLabels, err := s.getCurrentLabelsForRuntime(ctx, rtmTenant, runtimeID)
	if err != nil {
		return err
	}


	newRuntimeLabels :=  currentRuntimeLabels // TODO: copy it; this are labels after modification
	delete(newRuntimeLabels, key)

	err = s.upsertScenariosLabelIfShould(ctx, runtimeID, key, currentRuntimeLabels, newRuntimeLabels)
	if err != nil {
		return err
	}

	err = s.labelRepo.Delete(ctx, rtmTenant, model.RuntimeLabelableObject, runtimeID, key)
	if err != nil {
		return errors.Wrapf(err, "while deleting Runtime label")
	}

	return nil
}

func (s *service) ensureRuntimeExists(ctx context.Context, tnt string, runtimeID string) error {
	rtmExists, err := s.repo.Exists(ctx, tnt, runtimeID)
	if err != nil {
		return errors.Wrap(err, "while checking Runtime existence")
	}
	if !rtmExists {
		return fmt.Errorf("Runtime with ID %s doesn't exist", runtimeID)
	}

	return nil
}

func (s *service) upsertScenariosLabelIfShould(ctx context.Context, runtimeID string, modifiedLabelKey string, currentRuntimeLabels, newRuntimeLabels map[string]interface{}) error {
	rtmTenant, err := tenant.LoadFromContext(ctx)
	if err != nil {
		return errors.Wrapf(err, "while loading tenant from context")
	}

	finalScenarios := make([]interface{}, 0)

	if modifiedLabelKey == model.ScenariosKey {
		scenarios, err := s.scenarioAssignmentEngine.MergeScenariosFromInputAndAssignmentsFromInput(ctx, newRuntimeLabels)
		if err != nil {
			return errors.Wrap(err, "while merging scenarios from input and assignments")
		}

		for _, scenario := range scenarios {
			finalScenarios = append(finalScenarios, scenario)
		}
	} else {
		oldScenariosLabel, previousScenariosFromAssignments, err := s.getOldScenariosLabelAndPreviousScenariosFromAssignments(ctx, currentRuntimeLabels)
		if err != nil {
			return errors.Wrap(err, "while getting old scenarios label and scenarios from assignments")
		}

		newScenariosFromAssignments, err := s.getNewScenariosFromAssignments(ctx, newRuntimeLabels)
		if err != nil {
			return errors.Wrap(err, "while getting new scenarios from assignments")
		}

		finalScenarios = s.scenarioAssignmentEngine.ComputeScenarios(oldScenariosLabel, previousScenariosFromAssignments, newScenariosFromAssignments)
	}

	if len(finalScenarios) == 0 {
		return nil
	}

	scenariosLabelInput := &model.LabelInput{
		Key:        model.ScenariosKey,
		Value:      finalScenarios,
		ObjectID:   runtimeID,
		ObjectType: model.RuntimeLabelableObject,
	}

	err = s.labelUpsertService.UpsertLabel(ctx, rtmTenant, scenariosLabelInput)
	if err != nil {
		return errors.Wrapf(err, "while creating scenarios label for Runtime with id [%s]", runtimeID)
	}

	return nil
}
```

File: components/director/internal/domain/runtime/service.go [hunk 5]
```diff
@@ -278,10 +366,127 @@ func (s *service) DeleteLabel(ctx context.Context, runtimeID string, key string)
...
+	}
+
+	currentLabels := make(map[string]interface{})
+	for _, v := range labels {
+		currentLabels[v.Key] = v.Value
+	}
+	return currentLabels, nil
+}
+
+func (s *service) getOldScenariosLabelAndPreviousScenariosFromAssignments(ctx context.Context, currentRuntimeLabels map[string]interface{}) ([]interface{}, []interface{}, error) {
```
[the remark this answers is not in the document]
Comment (kfurgol) [reply components/director/internal/domain/runtime/service.go:438]: done

File: components/director/internal/domain/scenarioassignment/engine.go [hunk 6]
```diff
@@ -28,3 +39,123 @@ func (engine) RemoveAssignedScenarios(in []*model.AutomaticScenarioAssignment) e
...
+
+	scenarios := make([]string, 0)
+	for k := range scenariosSet {
+		scenarios = append(scenarios, k)
+	}
+
+	return scenarios, nil
+}
+
+func (e engine) MergeScenariosFromInputAndAssignmentsFromInput(ctx context.Context, inputLabels map[string]interface{}) ([]interface{}, error) {
```
[the remark this answers is not in the document]
Comment (kfurgol) [reply components/director/internal/domain/scenarioassignment/engine.go:76]: went with that name proposition

File: components/director/internal/domain/runtime/service.go [hunk 7]
```diff
@@ -270,18 +328,166 @@ func (s *service) DeleteLabel(ctx context.Context, runtimeID string, key string)
...
+			return errors.Wrap(err, "while getting new scenarios from assignments")
+		}
+
+		finalScenarios = s.scenarioAssignmentEngine.MergeScenarios(oldScenariosLabel, previousScenariosFromAssignments, newScenariosFromAssignments)
+	}
+
+	if len(finalScenarios) == 0 {
+		err := s.labelRepo.Delete(ctx, rtmTenant, model.RuntimeLabelableObject, runtimeID, model.ScenariosKey)
+		if err != nil {
+			return err
```
Comment (pkosiec) [inline components/director/internal/domain/runtime/service.go:414]: wrap error pls

**Pull request merged**